File: codes/2025.01/2024.01.21/1.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
var getDirections = function (root, startValue, destValue) {
    let startPath = []; // 根节点到起点
    let destPath = []; // 根节点到目的点

    // 找到两条路径
    findPath(root, startValue, startPath);
    findPath(root, destValue, destPath);

    function findPath(node, target, path) {
        if (!node) return false;
        // 找到目标节点
        if (node.val === target) {
            return true;
        }
        // 尝试左子树
        if (findPath(node.left, target, path)) {
            path.unshift("L"); // 路径是从节点到根的，使用 unshift
            return true;
        }
        // 尝试右子树
        if (findPath(node.right, target, path)) {
            path.unshift("R"); // 路径是从节点到根的，使用 unshift
            return true;
        }
        return false;
    }

    // 跳过公共前缀
    let i = 0;
    while (
        i < startPath.length &&
        i < destPath.length &&
        startPath[i] === destPath[i]
    ) {
        i++;
    }

    // 构造最终路径
    // 从起点到公共祖先需要向上走
    let result = "U".repeat(startPath.length - i);
    // 从公共祖先到终点需要按照destPath走
    result += destPath.slice(i).join("");

    return result;
};

File: codes/2024/2024.09/6/1.py
# 二分查找
def binary_search(list, target):
    left = 0
    right = len(list) - 1
    while left <= right:
        ## // 是整数除法运算符（向下取整除法）
        mid_index = (left + right) // 2
        mid = list[mid_index]
        # 如果目标值大于中间值，则更新右边界
        # 在左半边查找
        if target > mid:
            right = mid - 1
        # 如果目标值小于中间值，则更新左边界
        # 在右半边查找
        elif target < mid:
            left = mid + 1
        elif target == mid:
            return mid
    return None
